### ISSUE TITLE
Linear regression with two exactly collinear columns gives uninformative error

### ISSUE THREAD

**Issue by sebastianskejoe** [JuliaStats/GLM.jl#250]
Fitting a linear model to a dataset with two exactly collinear columns, gives the following (somewhat) unhelpful error message:
```
ERROR: PosDefException: matrix is not positive definite; Cholesky factorization failed.
Stacktrace:
...
```
It would be nice to provide information on which columns are collinear and/or suggest setting the third argument of `lm()` to `true` (currently undocumented) as suggested on [discourse](https://discourse.julialang.org/t/linear-regression-with-a-positive-definite-matrix-in-glm-jl/14638/9).

**Comment by pdeffebach** [comment]
I want to advocate more clearly for why this behavior should silently use missings. 

Our stata workflow looks something like this, for a paper on crime on city blocks in a city. 

```
global controls baseline_violent_crime baseline_property_crime  density income $district_dummies
global dependent_variables endline_violent_crime endline_property_crime

for y in $dependent_variables { // this is usually a function 
    /* In julia, this will have to be done with an `@eval` command 
        that constructs a new formula each time and allows it to be used in 
        `lm(). */
    regress `y' treatment $controls 
    // store the beta somewhere
    }
```

Now we want to do heterogeneity by baseline crime level. Though we have `baseline_violent_crime` and `baseline_property_crime` separately, we only want to do heterogeneity on both added together. 

```
gen baseline_crime_level = baseline_violent_crime + baseline_property_crme
gen treat_baseline_crime_level = treatment * baseline_crime_level // make the variable for summary stats
```

Now we have this new regression to run, but we still want to have our existing set of controls to use.

```
for y in $dependent_variables { // this is often a function 
    /* This would throw an error in Julia because baseline_crime_level
        is the sum of two variables in our control global */ 
    regress `y' treatment treatment_baseline_crime_level baseline_crime_level $controls
    // store the beta somewhere
    }
```

Ideally, any construction of variables, even interaction terms, are done in the "variable construction" scripts, and no new variables are created in the "analysis" script. This makes the problem difficult because if I see an interaction term in the analysis script, I *really* don't want to sift through our vector of controls every time, find the collinear variables and update the global so that they are not listed as controls. I also don't want to end up with a multitude of `$control` globals. I would rather set them in one central place. 

Perhaps I am thinking too much in a Stata mode of thinking here. If the mantainers of GLM could chime in and see how GLM and non-positive definite matrices would work with the above workflow, I would appreciate it.

**Comment by nalimilan** [comment]
@pdeffebach Does it work if you pass `true` as the third argument to `lm`? If not, better file a separate issue about that. Whether it should be the default is yet another question.

**Comment by pdeffebach** [comment]
Ah yes. Thanks for that. 

I'm fine with this, then. Given that this interaction scenario is not the main specification, I'm fine specifying `true` when needed. 

I can submit a PR for docs on this. Though perhaps it should be a kw for clarity.

**Comment by pdeffebach** [comment]
A few thoughts on this: 

You can't use `wts` option with `lm()`, but you can't say `AllowRankDeficient` with `glm()`

We should allow both.

**Comment by pdeffebach** [comment]
Regarding `AllowRankDeficient` I think the solution is in this function: 

https://github.com/JuliaStats/GLM.jl/blob/862106592b8d2c4664d25df355824c8df12977b6/src/glmfit.jl#L299

We need a keyword argument with a default value `false` to be passed to `cholpred()` in the following line: 

 `res = M(rr, cholpred(X), false)`

**Comment by nico202** [comment]
Maybe this has already been answered, but what to do if 
```julia
glm(@formula(resp ~ stim), data, Binomial(), ProbitLink())
```
fails with the 

```
 PosDefException: matrix is not positive definite; Cholesky factorization failed.
```
?

(R with the same data does not complain)

**Comment by nalimilan** [comment]
See https://github.com/JuliaStats/GLM.jl/pull/314.

**Comment by nico202** [comment]
Cool, thanks

**Comment by nalimilan** [comment]
Fixed by #386.
